### PR TITLE
 parallelized `update` + infrastructure for fast shell-based commands

### PR DESCRIPTION
**Pull request by geoff-codes**
Following up on discussion in #38818, this PR tracks a topic branch I've created for implementing a much faster `brew update` (the idea being that, once stable, it could/would run automatically).

I've still yet to implement the latest discussed changes in #38818, but part and parcel of this (and prerequisite to it) is some additional infrastructure for handling commands "outside" ruby.

At this time, this PR consists of:

- The new shell-based `update` command (WIP).

- Some necessary changes to the `bin/brew` shell script:
  - The script now fully dereferences its path(s). Previously, one level of symbolic links.
  - A switch to handle parsing and executing subcommands, **prior** to to the ruby `exec`
  - To illustrate and test, I've implemented `--prefix` in-line in the script, and the handling of both explicit (`update`) and implicit (i.e., from a tap, or elsewhere in one's path) subcommands.

- A small shell utility library, consisting of some helpful convenience functions I've written.
  `brew shellutils` both illustrates and documents its usage.

Additionally, I've put a couple of example commands in a tap, `geoff-codes/fastcommands`, to illustrate how all the above could be utilized.

I'll post some more benchmarks over the next couple of days, in addition to fleshing out the new update command. In the meantime, please feel free to kick the tires, give your thoughts or contribute!

:beers:

### CODE REVIEW

File: Library/Homebrew/cmd/update [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    { git stash save --include-untracked >/dev/null 2>&1 || :;} &&
+    { git pull --ff --no-rebase >/dev/null 2>&1 ;}              &&
+    { git stash pop >/dev/null 2>&1 || :;}
+
+    etag=$(echo "$response" | grep 'ETag:' | sed 's|.* "\(.*\)"|\1|')
+    git config brew.etag $(echo $etag | tr -d '\r') # Updated cached ETag
+  }
+)
+
+update_if_necessary "$HOMEBREW_PREFIX"
```
Comment (geoff-codes) [inline Library/Homebrew/cmd/update]: By separating this out and letting the taps update in the background, if we've nothing to update, we're down to about 0.65 seconds. For reference, `time curl -so /dev/null https://api.github.com` is 0.55 seconds from my IP.

File: Library/Homebrew/cmd/shellutils [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+}
+
+##===--------------------------------------------------------------------===##
+##===--- end helper functions -------------------------------------------===##
+##===--------------------------------------------------------------------===##
+
+case $(basename "$0") in
+   *shellutils) brew_shellutils_main "$@" ;;
+   *|'') continue ;;
+esac
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/shellutils]: I think we can do without the helper methods for a first past.

File: Library/Homebrew/cmd/update [hunk 3]
```diff
@@ -0,0 +1,54 @@
+#!/bin/sh # brew-update.sh
+[ -n "$HOMEBREW_VERBOSE" ] && set -x
```
Comment (geoff-codes) [inline Library/Homebrew/cmd/update]: This (that is, `set -x` being woefully deficient for logging/debugging, particularly with the backgrounded jobs) was in large part the reason I put together the shell utility library. You can look forward to something much prettier.

File: bin/brew [hunk 4]
```diff
@@ -1,42 +1,98 @@
...
-if [ -z "$HOMEBREW_DEVELOPER" ]
-then
-  unset HOMEBREW_RUBY_PATH
-fi
-
-if [ -z "$HOMEBREW_RUBY_PATH" ]
-then
-  if [ "$(uname -s)" = "Darwin" ]
-  then
-    HOMEBREW_RUBY_PATH="/System/Library/Frameworks/Ruby.framework/Versions/Current/usr/bin/ruby"
```
Comment (geoff-codes) [inline bin/brew]: Is there a reason we can't use `/usr/bin/ruby`?

File: Library/Homebrew/cmd/update [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+trap_cleanup()
+{
+  for job in $(jobs -p); do
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update]: I'm not super familiar with jobs so am probably wrong but: could this not affect other jobs the user might have running in their environment?

File: Library/Homebrew/cmd/update [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    git stash pop >/dev/null 2>&1
+  done
+}
+
+trap trap_cleanup HUP INT QUIT ABRT KILL TERM
+
+update_if_necessary()
+(
+  cd "$*"
+  trap trap_cleanup HUP INT QUIT ABRT KILL TERM
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update]: Probably just worth trapping `INT` and maybe `HUP`.

File: Library/Homebrew/cmd/update [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  # Build the URI for the API call based on our remote and branch.
+  uri=$(git ls-remote --get-url | sed -e's|git://|https://|' -e's|\.git$||' \
+                                      -e's|github.com|api.github.com/repos|')
+  ref=$(git symbolic-ref HEAD)
+  api_uri=$(echo $uri | sed 's|$|/git|')/$ref
+
+  # We save a few milliseconds with a standard config avoiding the refs API.
+  [ "$ref" = "refs/heads/master" ] && api_uri="$uri/commits/master"
+
+  etag=$(git config brew.etag) # ETag gets cached in .git/config.
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update]: This config value should probably vary for each of the taps.

File: Library/Homebrew/cmd/shellutils [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+echo(){   /bin/echo "$@"              ;}
+print(){  /bin/echo "$@"              ;}
+printf(){ /bin/echo "$@" | tr -d '\n' ;}
+
+##===--------------------------------------------------------------------===##
+
+# Use `tput` for safe terminal colors, so we don't have to hardcode escapes.
+# `tput` is a POSIX utility and outputs proper escapes based on terminfo.
+# This is much safer than hardcoding escape sequences.
+
```
Comment (geoff-codes) [inline Library/Homebrew/cmd/shellutils]: To see how this degrades properly: `brew shellutils | cat`.

File: Library/Homebrew/cmd/update (hunk 1 above)
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update]: Is it worth making this in parallel too and then doing a `wait` for all the jobs to finish?

File: Library/Homebrew/cmd/shellutils (hunk 2 above)
Comment (geoff-codes) [inline Library/Homebrew/cmd/shellutils]: I can put it in the tap, move what I need into `update`.

File: bin/brew [hunk 9]
```diff
@@ -1,42 +1,98 @@
...
-  else
-    HOMEBREW_RUBY_PATH="$(which ruby)"
-  fi
-fi
```
Comment (mikemcquaid) [inline bin/brew]: Similarly let's remove all the changes to this file for now except for a special-case for `update`.

File: Library/Homebrew/cmd/update [hunk 10]
```diff
@@ -0,0 +1,54 @@
+#!/bin/sh # brew-update.sh
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update]: Maybe go for `update.sh` to make it more obvious the difference between the shell and Ruby versions.

File: bin/brew (hunk 9 above)
Comment (mikemcquaid) [inline bin/brew]: Particularly any refactoring which makes it hard to follow.

File: Library/Homebrew/cmd/update (hunk 5 above)
Comment (geoff-codes) [inline Library/Homebrew/cmd/update]: Uh... good question, but I just double-checked; no. `jobs -p` are only processes spawned by the parent process.

File: Library/Homebrew/cmd/update (hunk 6 above)
Comment (geoff-codes) [inline Library/Homebrew/cmd/update]: Yeah.

File: Library/Homebrew/cmd/update (hunk 7 above)
Comment (geoff-codes) [inline Library/Homebrew/cmd/update]: And by branch and remote, probably.

Comment (mikemcquaid) [inline Library/Homebrew/cmd/update]: Well, this path is going to only be used to get the `master` branch on `origin`.

File: Library/Homebrew/cmd/shellutils (hunk 2 above)
Comment (mikemcquaid) [inline Library/Homebrew/cmd/shellutils]: I'd rather we just tried to do it in raw bash and worry about making it look pretty at a later pass.

File: Library/Homebrew/cmd/update (hunk 1 above)
Comment (geoff-codes) [inline Library/Homebrew/cmd/update]: So, what's happening here is we're waiting for _homebrew_ (only _homebrew_) to finish updating; then spawning a process for each tap, which run concurrently; and silently complete and exit, while we can move on. Beyond that, there's not much more to parallelize; and while we could `wait` for all the taps to finish updating, but that could be a few seconds.

Its kinda like this:
```sh
#!/bin/sh
a()( echo "Hi, I'm subshell A, and I complete before you move on."; sleep 2 )
b()( while :; do echo "Hi, I'm subshell B."; sleep 1; done )
c()( while :; do echo "Hi, I'm subshell C."; sleep 1; done )
d()( while :; do echo "Hi, I'm subshell D."; sleep 1; done )
e()( while :; do echo "Hi, I'm subshell E."; sleep 1; done )
f()( while :; do echo "Hi, I'm subshell F."; sleep 1; done )
a
echo "Ok, thanks A. Now here these background subshells, but I'm not waiting!"
b & c & d & e & f & ## No waiting, onwards...
sleep 1
echo "Hey, this is the parent shell, and I'm gonna exec you, ruby!" # And they keep going after we `exec`.
exec ruby -e "sleep 1; puts 'Hey, ruby here, and I own these jobs now; time to go!'; sleep 1; %x(kill 0)"
```

File: bin/brew (hunk 9 above)
Comment (geoff-codes) [inline bin/brew]: Okiedoke.

File: Library/Homebrew/cmd/update (hunk 7 above)
Comment (geoff-codes) [inline Library/Homebrew/cmd/update]: So, I misread your first note there; these get stored each taps' own .git/config already.
And what I meant to say is that this should be something like `git config remote.origin.brew-api-tag` so that it is associated with [ remote "origin" ].

Comment (mikemcquaid) [inline Library/Homebrew/cmd/update]: :cool:

**Comment by mikemcquaid** [comment]
@geoff-codes Just in interest of full-disclosure: I'm working on my own branch for this and I'll happily credit your work but I suspect my PR will be the one merged rather than this one. Please feel free to keep updating this PR if you like though and we can collaborate. Thanks!

**Comment by DomT4** [comment]
@mikemcquaid Is this okay to close?

**Comment by geoff-codes** [comment]
@DomT4 Its fine by me. And I'm glad you were able to take the ball and run with it Mike, I've been swamped of late and hadn't had the time to hack on it anyway. Cheers!

**Comment by mikemcquaid** [comment]
Thanks for understanding @geoff-codes :bow:
